Improve RDB to AOF conversion docs (#2521)
redis/redis-doc: docs/management/persistence.md

### How I can switch to AOF, if I'm currently using dump.rdb snapshots?

> If you want to enable AOF in a server that is currently using RDB snapshots, you need to convert the data interactively using `redis-cli`.

Author: I assume you mean this comment? https://github.com/redis/redis/pull/12506#discussion_r1300149787 I amended the commit!

**IMPORTANT:** not following this procedure (e.g. just changing the config and restarting the server) can result in data loss!

**Redis >= 2.2**

* Make a backup of your latest dump.rdb file.
* Transfer this backup to a safe place.
* Enable AOF: `redis-cli config set appendonly yes`
* Optionally disable RDB: `redis-cli config set save ""`

> * Make sure your database contains the same number of keys it contained previously.

Reviewer: IMHO that line is excessive (and silly)

Author: I simply kept that line, which was already part of the docs before.

In general, a quick check that your data is still fine isn't totally wrong. Maybe we could word it like this?

    Make sure your database still contains all data, e.g. `redis-cli keys '*'`

Reviewer: but it's silly since enabling AOF can't (technically) affect your data in redis.
maybe instead we can add a line that suggests to wait till that initial AOFRW process completes?
i.e. if you'll try to restart redis before that happens, it'll refuse to go down.

Reviewer: sadly, i don't see that we exposed the status of `server.aof_state` in INFO, so we may need to suggest to look at some combination of `aof_rewrite_in_progress`, `aof_rewrite_scheduled`, and maybe `aof_last_bgrewrite_status`.
WDYT?

Author: > but it's silly since enabling AOF can't (technically) affect your data in redis.

That's what I thought, and then all data was gone after restarting Redis (and it took 4 hours for someone to notice) 😉 (I know what you mean though, from a technical point of view.)

> sadly, i don't see that we exposed the status of `server.aof_state` in INFO, so we may need to suggest to look at some combination of `aof_rewrite_in_progress`, `aof_rewrite_scheduled`, and maybe `aof_last_bgrewrite_status`. WDYT?

These are all keys that are listed when querying `INFO persistence`, correct? What would be the correct combination of keys? (I don't know enough about Redis internals to propose some wording.)

Reviewer: > > but it's silly since enabling AOF can't (technically) affect your data in redis.
> 
> That's what I thought, and then all data was gone after restarting Redis (and it took 4 hours for someone to notice) 😉 (I know what you mean though, from a technical point of view.)

but that line tells you to check the data before restart, when all you did is change a configuration that doesn't affect the data in any way (redis doesn't touch it), so it's silly.
i.e. it does affect the guarantees, and how that data is persisted, but that's not what that line suggests to check.
IMHO we should delete it.

> > sadly, i don't see that we exposed the status of `server.aof_state` in INFO, so we may need to suggest to look at some combination of `aof_rewrite_in_progress`, `aof_rewrite_scheduled`, and maybe `aof_last_bgrewrite_status`. WDYT?
> 
> These are all keys that are listed when querying `INFO persistence`, correct? What would be the correct combination of keys? (I don't know enough about Redis internals to propose some wording.)

how about:
```
You may want to wait for AOF rewrite to finish persisting the data. You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validate that `aof_last_bgrewrite_status` is `ok`.
```

Reviewer 2: This (check that the database contains the same number of keys it contained previously) makes sense to do after restarting. That's where things can go wrong.

I think it is better to move this bullet to a step after all other steps "Next time you restart the server, check that..."

I think we shall also add Oran's suggestion. It is important especially if you want to restart redis just after changing the config, when AOF rewrite is ongoing. I think we should also add "Before restarting, " to Oran's suggestion, so:

> Before restarting, you may want to wait for AOF rewrite to finish persisting the data. You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validate that `aof_last_bgrewrite_status` is `ok`.

* Make sure writes are appended to the append only file correctly.

> * Update your `redis.conf` and ensure that it matches the config above.

Reviewer: maybe add a reference to CONFIG REWRITE?

Author: Maybe like this?

> * Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the config above.

> * Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the config above.

Reviewer 2: A few lines below this, there is:

> **IMPORTANT:** remember to edit your redis.conf to turn on the AOF, otherwise
when you restart the server the configuration changes will be lost and the

We should combine these two since they're saying the same thing. Can we move the IMPORTANT into this bullet?

Author: Yes, good point. I think the new structure (Preparations / Switch / After Restart) helps with that as well.

### How I can switch to AOF, if I'm currently using dump.rdb snapshots?

If you want to enable AOF in a server that is currently using RDB snapshots, you need to convert the data by enabling AOF via CONFIG command on the live server first.

**IMPORTANT:** not following this procedure (e.g. just changing the config and restarting the server) can result in data loss!

**Redis >= 2.2**

Preparations:

* Make a backup of your latest dump.rdb file.
* Transfer this backup to a safe place.

Switch to AOF on live database:

* Enable AOF: `redis-cli config set appendonly yes`
* Optionally disable RDB: `redis-cli config set save ""`
* Make sure writes are appended to the append only file correctly.
* **IMPORTANT:** Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the config above.
  (If you forget this step, when you restart the server, the configuration changes will be lost and the server will start again with the old configuration, resulting in a loss of your data.)

Next time you restart the server:

> * Wait for AOF rewrite to finish persisting the data.

Reviewer 2: ```suggestion
* Before restarting the server, wait for AOF rewrite to finish persisting the data.
```

  You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validate that `aof_last_bgrewrite_status` is `ok`.

> * Make sure your database contains the same number of keys it contained previously.

Reviewer 2: ```suggestion
* After restarting the server, check that your database contains the same number of keys it contained previously.
```

### How I can switch to AOF, if I'm currently using dump.rdb snapshots?

If you want to enable AOF in a server that is currently using RDB snapshots, you need to convert the data by enabling AOF via CONFIG command on the live server first.

**IMPORTANT:** not following this procedure (e.g. just changing the config and restarting the server) can result in data loss!

**Redis >= 2.2**

Preparations:

* Make a backup of your latest dump.rdb file.
* Transfer this backup to a safe place.

Switch to AOF on live database:

* Enable AOF: `redis-cli config set appendonly yes`
* Optionally disable RDB: `redis-cli config set save ""`

> * Wait for AOF rewrite to finish persisting the data. You can do that by

Reviewer 2: This is important to check only if you want to restart the server. It is not necessary to restart the server. Maybe move this to "Before restarting the server" section, or at least mention that it's related to restarting?

Author: Good point, updated in 2f202c889f73918a2f377737b015d287ddf3debd

  watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and
  `aof_rewrite_scheduled` to be `0`, and validate that
  `aof_last_bgrewrite_status` is `ok`.
* **IMPORTANT:** Update your `redis.conf` (potentially through `CONFIG
  REWRITE`) and ensure that it matches the config above. (If you forget this
  step, when you restart the server, the configuration changes will be lost and
  the server will start again with the old configuration, resulting in a loss
  of your data.)

Next time you restart the server:

* Make sure your database contains the same number of keys it contained previously.

> * Make sure writes are appended to the append only file correctly.

Reviewer 2: Writes are appended to AOF file as soon as AOF rewrite is finished. Restart is not necessary. It is more related to the bullet "Wait for AOF rewrite to finish ... INFO persistence ..." above, so maybe add this in the end of that bullet?

### How I can switch to AOF, if I'm currently using dump.rdb snapshots?

If you want to enable AOF in a server that is currently using RDB snapshots,
you need to convert the data interactively using `redis-cli`.

**IMPORTANT:** not following this procedure (e.g. just changing the config and
restarting the server) can result in data loss!

The following steps require at least Redis 2.2:

* Make a backup of your latest dump.rdb file.
* Transfer this backup to a safe place.
* Enable AOF: `redis-cli config set appendonly yes`
* Optionally disable RDB: `redis-cli config set save ""`
* Make sure your database contains the same number of keys it contained
  previously.
* Make sure writes are appended to the append only file correctly.
* Update your `redis.conf` and ensure that it matches the config above.

The first CONFIG command enables the Append Only File persistence.

The second CONFIG command is used to turn off snapshotting persistence. This is

> optional, if you wish you can take both the persistence methods enabled.

Reviewer 3: We should avoid this change that just adding an extra line.

Author: I did that for consistency with the rest of the file, but I can revert the change.

* Make sure writes are appended to the append only file correctly.
* Update your `redis.conf` and ensure that it matches the config above.

The first CONFIG command enables the Append Only File persistence.

The second CONFIG command is used to turn off snapshotting persistence. This is optional, if you wish you can take both the persistence methods enabled.
The second CONFIG command is used to turn off snapshotting persistence. This is
optional, if you wish you can take both the persistence methods enabled.

**IMPORTANT:** remember to edit your redis.conf to turn on the AOF, otherwise
when you restart the server the configuration changes will be lost and the
server will start again with the old configuration.

**Redis 2.0**

* Make a backup of your latest dump.rdb file.
* Transfer this backup into a safe place.
* Stop all the writes against the database!
* Issue a `redis-cli BGREWRITEAOF`. This will create the append only file.
* Stop the server when Redis finished generating the AOF dump.
* Edit redis.conf end enable append only file persistence.
* Restart the server.
* Make sure that your database contains the same number of keys it contained before the switch.

> * Make sure that writes are appended to the append only file correctly.

Reviewer 3: Delete by mistake?

Reviewer 3: Not really, I still see some users still using redis 3.x.

Author: Redis 3 was released 4 years after 2.0, and the current instructions will work from 2.2 onwards. But sure, I can re-add these instructions.

* Enable AOF: `redis-cli config set appendonly yes`
* Optionally disable RDB: `redis-cli config set save ""`
* Make sure writes are appended to the append only file correctly.

> * **IMPORTANT:** Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the config above.

Reviewer 4: ```suggestion
* **IMPORTANT:** Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the configuration above.
```

>   (If you forget this step, when you restart the server, the configuration changes will be lost and the server will start again with the old configuration, resulting in a loss of your data.)

Reviewer 4: ```suggestion
  If you forget this step, when you restart the server, the configuration changes will be lost and the server will start again with the old configuration, resulting in a loss of your data.
```

Next time you restart the server:

* Wait for AOF rewrite to finish persisting the data.

>   You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validate that `aof_last_bgrewrite_status` is `ok`.

Reviewer 4: ```suggestion
  You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validating that `aof_last_bgrewrite_status` is `ok`.
```
